api/distributor/distributor.proto

```diff
@@ -0,0 +1,33 @@
+syntax = "proto3";
+
+package nighthawk;
+
+import "envoy/config/core/v3/address.proto";
+
+import "api/client/service.proto";
+import "api/sink/sink.proto";
+
+import "validate/validate.proto";
+
+// Perform an execution request through an intermediate service that will in
+// turn delegate to one or more other services for actual handling.
+message DistributedRequest {
+  oneof distributed_request_type {
+    client.ExecutionRequest execution_request = 1;
```

Comment:
1. `DistributedResponse` is empty in this PR, its contents will be added in follow up(s).
2. When a `DistributedRequest` wraps a `SinkRequests`, the number of `services` *must* equal precisely one. I can see the confusion, but it doesn't seem to fit into the declarative validation model to annotate this as such. 

Some thoughts:

Initially the idea was to have the distributor be a "generic" service which could be described as one that would accept arbitrary messages and forward them as-is to 1..n services within the cluster, and subsequently stream back the replies (wrapping these reply messages to annotate them with which internal service was associated to replies on the stream). Consensus was that it would be better to not attempt the generic approach and instead be specific, hence the explicit wrapper messages we use for distributed requests & replies. With that, the possibility to address > 1 sinks through the distributor stands out, and I think that is what is causing confusion.  I'd be up for iterating on suggestions to enhance the modelling here. The `services` field is shared across the message types that can be distributed, but the constraints on this field varies across message types, and right now this "knowledge" is implicit in clients consuming the distributor service. 

For clarity and ensuring we're all on the same page, here's a diagram of what this is converging to:

![image](https://user-images.githubusercontent.com/2239556/111592911-afa5ff80-87c9-11eb-8c24-207d304f2e62.png)

Reply:
Thank you for the explanation @oschaaf. I am assuming that when `SinkRequest` is sent, the client is trying to consume data stored in the sink service, rather than write them. With that I would like us to consider two suggestions that would remove this confusion.

1. Does the client need to access the Sink service via the Distributor service, or are they chained just for convenience? Assuming the Distributor doesn't provide any critical functionality to the client - there is nothing wrong with dialing the Sink from the client directly. Would that work?
2. If we do identify critical functionality that the Distributor needs to provide to the client when retrieving the results, I would suggest we make the approach more specific by defining a second RPC method solely for the purposes of getting the results from the Sink.

Reply:
It is a convenience thing. Option 1. would be low impact to move to, I will lean the api in this PR up.

Reply:
Done in 22060358

Reply:
Thanks for updating here Otto!

Reply:
Extended alternate/diagram that includes the adaptive load controller:

![image](https://user-images.githubusercontent.com/2239556/111705929-bc663a00-8841-11eb-9560-0951e4a688ed.png)

In this flow the adaptive load controller runs behind nighthawk_service, which avoids the need to teach the adaptive load controller about sink and distributor services. (nighthawk_service will do that on its behalf).